contracts/interchain-token-service/tests/flow_limit.rs

```diff
@@ -0,0 +1,290 @@
+    approve_gateway_messages(&env, &gateway_client, signers, messages);
+
+    client.execute(&source_chain, &message_id, &source_address, &payload);
+}
+
+#[test]
+fn add_flow_out_succeeds() {
+    let (env, client, _, _, _) = setup_env();
+    register_chains(&env, &client);
+    let (token_id, sender) = setup_flow_limit(&env, &client);
+    let gas_token = setup_gas_token(&env, &sender);
+
+    let amount = 1000;
+    let destination_chain = String::from_str(&env, "ethereum");
+    let destination_address = Bytes::from_hex(&env, "4F4495243837681061C4743b74B3eEdf548D56A5");
+    let data = None;
+
+    client
+        .mock_all_auths()
+        .set_trusted_chain(&destination_chain);
+
+    client.mock_all_auths().interchain_transfer(
+        &sender,
+        &token_id,
+        &destination_chain,
+        &destination_address,
+        &amount,
+        &data,
+        &gas_token,
+    );
+
+    assert_eq!(client.flow_out_amount(&token_id), amount);
+}
+
+#[test]
+#[should_panic(expected = "Error(Contract, #20)")] // ContractError::FlowLimitExceeded
+fn add_flow_out_fails_exceeds_flow_limit() {
+    let (env, client, _, _, _) = setup_env();
+    register_chains(&env, &client);
+    let (token_id, sender) = setup_flow_limit(&env, &client);
+    let gas_token = setup_gas_token(&env, &sender);
+
+    let amount = TEST_FLOW_LIMIT.unwrap();
+    let destination_chain = String::from_str(&env, "ethereum");
+    let destination_address = Bytes::from_hex(&env, "4F4495243837681061C4743b74B3eEdf548D56A5");
+    let data = None;
+
+    client
+        .mock_all_auths()
+        .set_trusted_chain(&destination_chain);
+
+    client.mock_all_auths().interchain_transfer(
+        &sender,
+        &token_id,
+        &destination_chain,
+        &destination_address,
+        &amount,
+        &data,
+        &gas_token,
+    );
+
+    let second_amount = 1;
+
+    client.mock_all_auths().interchain_transfer(
+        &sender,
+        &token_id,
+        &destination_chain,
+        &destination_address,
+        &second_amount,
+        &data,
+        &gas_token,
+    );
+}
+
+#[test]
+#[should_panic(expected = "Error(Contract, #21)")] // ContractError::FlowAmountOverflow
+fn add_flow_fails_on_flow_comparison_overflow() {
+    let (env, client, gateway_client, _, signers) = setup_env();
+    register_chains(&env, &client);
+    let (token_id, sender) = setup_flow_limit(&env, &client);
+    let gas_token = setup_gas_token(&env, &sender);
+
+    client
+        .mock_all_auths()
+        .set_flow_limit(&token_id, &Some(i128::MAX - 50));
+
+    let high_amount = i128::MAX - 100;
+    let (source_chain, message_id, source_address, payload, messages) =
+        create_interchain_transfer_message(&env, &client, &token_id, high_amount);
+    approve_gateway_messages(&env, &gateway_client, signers, messages);
+    client.execute(&source_chain, &message_id, &source_address, &payload);
+
+    let small_amount = 100;
```

Comment:
```suggestion
    let small_amount = 2;
```

Reply:
I recommend creating a vector of flow in + flow out test cases, and running the test logic here in a loop over them to try different scenarios